[BUG] Endless WRITE_TXN_MARKERS requests even after producer is closed

**To Reproduce**

Run following application against a KoP standalone:

```java
        final var bootstrapServers = "localhost:9092";
        final var topic = "my-topic";

        @Cleanup final var admin = AdminClient.create(Collections.singletonMap(
                AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers));

        admin.createTopics(Collections.singleton(new NewTopic(topic, 1, (short) 1))).all().get();

        final var props = new Properties();
        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
        props.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, "X0");
        final var producer = new KafkaProducer<String, String>(props);
        producer.initTransactions();
        producer.beginTransaction();
        producer.send(new ProducerRecord<>(topic, "hello"));
        producer.commitTransaction();
        producer.close();

        admin.deleteTopics(Collections.singleton(topic)).all().get();
```

- Kafka Client: 3.1.0
- Pulsar: 2.10.0
- KoP: 2.10.1.1

KoP configs:

```properties
messagingProtocols=kafka
protocolHandlerDirectory=./protocols
allowAutoTopicCreationType=partitioned
kafkaListeners=PLAINTEXT://127.0.0.1:9092
brokerEntryMetadataInterceptors=org.apache.pulsar.common.intercept.AppendIndexMetadataInterceptor
brokerDeleteInactiveTopicsEnabled=false
brokerDeleteInactivePartitionedTopicMetadataEnabled=true
brokerDeduplicationEnabled=true
kafkaTransactionCoordinatorEnabled=true
brokerDeleteInactiveTopicsMode=delete_when_subscriptions_caught_up
```

**Expected behavior**
A clear and concise description of what you expected to happen.

**Additional context**

[completed-logs.log](https://github.com/streamnative/kop/files/9052910/pulsar-standalone-xuyunzedeMacBook-Pro.local.log)

If we slept for a while before deleting the topic, the issue would become the same with https://github.com/streamnative/kop/issues/1387.

```java
        producer.close();
        Thread.sleep(1000);
        admin.deleteTopics(Collections.singleton(topic)).all().get();
```

```
Exception in thread "main" java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.UnknownTopicOrPartitionException: This server does not host this topic-partition.
Caused by: org.apache.kafka.common.errors.UnknownTopicOrPartitionException: This server does not host this topic-partition.
```

--- a/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopBrokerLookupManager.java
+++ b/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopBrokerLookupManager.java
@@ -27,6 +27,8 @@ import lombok.extern.slf4j.Slf4j;
 import org.apache.commons.lang3.StringUtils;
 import org.apache.pulsar.broker.PulsarService;
 import org.apache.pulsar.broker.resources.MetadataStoreCacheLoader;
+import org.apache.pulsar.client.admin.PulsarAdmin;
+import org.apache.pulsar.client.admin.PulsarAdminException;
 import org.apache.pulsar.common.naming.TopicName;
 import org.apache.pulsar.policies.data.loadbalancer.LoadManagerReport;
 import org.apache.pulsar.policies.data.loadbalancer.ServiceLookupData;
@@ -42,6 +44,8 @@ public class KopBrokerLookupManager {
     private final LookupClient lookupClient;
     private final MetadataStoreCacheLoader metadataStoreCacheLoader;
 
+    private final PulsarAdmin adminClient;
+
     private final AtomicBoolean closed = new AtomicBoolean(false);
 
     public static final ConcurrentHashMap<String, ConcurrentHashMap<String, CompletableFuture<InetSocketAddress>>>
@@ -52,6 +56,7 @@ public class KopBrokerLookupManager {
 
     public KopBrokerLookupManager(KafkaServiceConfiguration conf, PulsarService pulsarService) throws Exception {
         this.advertisedListeners = conf.getKafkaAdvertisedListeners();
+        this.adminClient = pulsarService.getAdminClient();
         this.lookupClient = KafkaProtocolHandler.getLookupClient(pulsarService);
         this.metadataStoreCacheLoader = new MetadataStoreCacheLoader(pulsarService.getPulsarResources(),
                 conf.getBrokerLookupTimeoutMs());
@@ -140,6 +145,23 @@ public class KopBrokerLookupManager {
             return CompletableFuture.completedFuture(null);
         }
         return lookupClient.getBrokerAddress(TopicName.get(topic), listenerName);
+    }
+
+    public CompletableFuture<Boolean> isTopicExists(final String topic) {
+        CompletableFuture<Boolean> future = new CompletableFuture<>();
+        this.adminClient.topics().getPartitionedTopicMetadataAsync(topic)
+                .whenComplete((__, ex) -> {
+                    if (ex != null) {
+                        if (ex instanceof PulsarAdminException.NotFoundException) {
+                            future.complete(false);
+                            return;
+                        }
+                        // Retry when the exception is others exception.
+                        log.error("Get partitioned topic metadata has exception.", ex);
+                    }
+                    future.complete(true);
+                });
+        return future;
     }
 
     private CompletableFuture<Optional<String>> getProtocolDataToAdvertise(

--- a/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMarkerChannelHandler.java
+++ b/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMarkerChannelHandler.java
@@ -131,7 +131,7 @@ public class TransactionMarkerChannelHandler extends ChannelInboundHandlerAdapte
         }
 
         public void onError(Throwable error) {
-            log.info("[TransactionMarkerChannelHandler] onError {}", error);
+            log.info("[TransactionMarkerChannelHandler] onError", error);
             final List<WriteTxnMarkersRequest.TxnMarkerEntry> markers = request.markers();
             Map<Long, Map<TopicPartition, Errors>> errors = new HashMap<>(markers.size());
             for (WriteTxnMarkersRequest.TxnMarkerEntry entry : markers) {
@@ -151,9 +151,6 @@ public class TransactionMarkerChannelHandler extends ChannelInboundHandlerAdapte
 
     @Override
     public void channelActive(ChannelHandlerContext channelHandlerContext) throws Exception {
-        if (log.isDebugEnabled()) {
-            log.debug("channelActive");
-        }
         log.info("[TransactionMarkerChannelHandler] channelActive to {}", channelHandlerContext.channel());
         handleAuthentication(channelHandlerContext);
         super.channelActive(channelHandlerContext);

--- a/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMarkerChannelManager.java
+++ b/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMarkerChannelManager.java
@@ -13,6 +13,7 @@
  */
 package io.streamnative.pulsar.handlers.kop.coordinator.transaction;
 
+import com.google.common.annotations.VisibleForTesting;
 import io.netty.bootstrap.Bootstrap;
 import io.netty.channel.EventLoopGroup;
 import io.netty.channel.nio.NioEventLoopGroup;
@@ -53,7 +54,6 @@ import org.apache.kafka.common.requests.WriteTxnMarkersRequest;
 import org.apache.kafka.common.requests.WriteTxnMarkersRequest.TxnMarkerEntry;
 import org.apache.pulsar.client.api.Authentication;
 import org.apache.pulsar.client.impl.AuthenticationUtil;
-import org.apache.pulsar.common.naming.TopicName;
 import org.apache.pulsar.common.util.FutureUtil;
 import org.apache.pulsar.common.util.netty.ChannelFutures;
 import org.eclipse.jetty.util.BlockingArrayQueue;
@@ -80,6 +80,9 @@ public class TransactionMarkerChannelManager {
     private Map<InetSocketAddress, CompletableFuture<TransactionMarkerChannelHandler>> handlerMap = new HashMap<>();
 
     private TransactionStateManager txnStateManager;
+
+    @Getter
+    @VisibleForTesting
     private ConcurrentHashMap<String, PendingCompleteTxn> transactionsWithPendingMarkers = new ConcurrentHashMap<>();
     private Map<InetSocketAddress, TxnMarkerQueue> markersQueuePerBroker = new ConcurrentHashMap<>();
     private TxnMarkerQueue markersQueueForUnknownBroker = new TxnMarkerQueue(null);
@@ -93,8 +96,9 @@ public class TransactionMarkerChannelManager {
     private Authentication authentication;
 
     @AllArgsConstructor
+    @Getter
     @ToString
-    private static class PendingCompleteTxn {
+    protected static class PendingCompleteTxn {
         private final String transactionalId;
         private final Integer coordinatorEpoch;
         private final TransactionMetadata txnMetadata;
@@ -245,11 +249,12 @@ public class TransactionMarkerChannelManager {
 
     public void maybeWriteTxnCompletion(String transactionalId) {
         ensureDrainQueuedTransactionMarkersActivity();
-        PendingCompleteTxn pendingCompleteTxn = transactionsWithPendingMarkers.get(transactionalId);
-        if (!hasPendingMarkersToWrite(pendingCompleteTxn.txnMetadata)
-                && transactionsWithPendingMarkers.remove(transactionalId, pendingCompleteTxn)) {
-            writeTxnCompletion(pendingCompleteTxn);
-        }
+        Optional.ofNullable(transactionsWithPendingMarkers.get(transactionalId)).ifPresent(pendingCompleteTxn -> {
+            if (!hasPendingMarkersToWrite(pendingCompleteTxn.txnMetadata)
+                    && transactionsWithPendingMarkers.remove(transactionalId, pendingCompleteTxn)) {
+                writeTxnCompletion(pendingCompleteTxn);
+            }
+        });
     }
 
     public void addTxnMarkersToBrokerQueue(String transactionalId,
@@ -269,32 +274,80 @@ public class TransactionMarkerChannelManager {
         for (TopicPartition topicPartition : topicPartitions) {
             String pulsarTopic = new KopTopic(topicPartition.topic(), namespacePrefixForUserTopics)
                     .getPartitionName(topicPartition.partition());
-            CompletableFuture<Optional<InetSocketAddress>> addressFuture =
-                    kopBrokerLookupManager.findBroker(TopicName.get(pulsarTopic), sslEndPoint);
             CompletableFuture<Void> addFuture = new CompletableFuture<>();
             addressFutureList.add(addFuture);
-            addressFuture.whenComplete((address, throwable) -> {
-                if (throwable != null) {
-                    log.warn("Failed to find broker for topic partition {}", topicPartition, throwable);
-                    unknownBrokerTopicList.add(topicPartition);
-                    addFuture.completeExceptionally(throwable);
-                    return;
-                }
-                if (!address.isPresent()) {
-                    log.warn("No address for broker for topic partition {}", topicPartition);
-                    unknownBrokerTopicList.add(topicPartition);
-                    addFuture.completeExceptionally(new Exception("no address for owner of " + topicPartition));
-                    return;
-                }
-                addressAndPartitionMap.compute(address.get(), (__, set) -> {
-                    if (set == null) {
-                        set = new ArrayList<>();
-                    }
-                    set.add(topicPartition);
-                    return set;
-                });
-                addFuture.complete(null);
-            });
+            kopBrokerLookupManager.isTopicExists(pulsarTopic)
+                    .thenAccept(isTopicExists -> {
+                        if (!isTopicExists) {
+                            ErrorsAndData<Optional<TransactionStateManager.CoordinatorEpochAndTxnMetadata>>
+                                    transactionState = txnStateManager.getTransactionState(transactionalId);
+                            if (transactionState.hasErrors()) {
+                                log.info("Encountered {} trying to fetch transaction metadata for {} with coordinator "
+                                                + "epoch {}; cancel sending markers to its partition leaders"
+                                , transactionState.getErrors(), transactionalId, coordinatorEpoch);
+                                transactionsWithPendingMarkers.remove(transactionalId);
+                                addFuture.complete(null);
+                                return;
+                            }
+                            Optional<TransactionStateManager.CoordinatorEpochAndTxnMetadata> epochAndTxnMetadata =
+                                    transactionState.getData();
+                            if (epochAndTxnMetadata.isPresent()) {
+                                if (!coordinatorEpoch.equals(epochAndTxnMetadata.get().getCoordinatorEpoch())) {
+                                    log.info("The cached metadata has changed to {} (old coordinator epoch is {}) "
+                                                    + "since preparing to send markers; "
+                                                    + "cancel sending markers to its partition leaders",
+                                            epochAndTxnMetadata, coordinatorEpoch);
+                                    transactionsWithPendingMarkers.remove(transactionalId);
+                                } else {
+                                    log.info("Couldn't find leader endpoint for partitions {} while trying "
+                                                    + "to send transaction markers for {}, these partitions are "
+                                                    + "likely deleted already and hence can be skipped",
+                                            topicPartition, transactionalId);
+                                    TransactionMetadata txnMetadata =
+                                            epochAndTxnMetadata.get().getTransactionMetadata();
+                                    txnMetadata.inLock(() -> {
+                                        topicPartitions.forEach(txnMetadata::removePartition);
+                                        return null;
+                                    });
+                                    maybeWriteTxnCompletion(transactionalId);
+                                }
+                            } else {
+                                log.error("The coordinator still owns "
+                                        + "the transaction partition for {}, but there is no metadata in the cache; "
+                                        + "this is not expected", transactionalId);
+                                addFuture.complete(null);
+                                return;
+                            }
+                            addFuture.complete(null);
+                            return;
+                        }
+
+                        CompletableFuture<Optional<InetSocketAddress>> addressFuture =
+                                kopBrokerLookupManager.findBroker(pulsarTopic, sslEndPoint);
+
+                        addressFuture.whenComplete((address, throwable) -> {
+                            if (throwable != null) {
+                                log.warn("Failed to find broker for topic partition {}", topicPartition, throwable);
+                                unknownBrokerTopicList.add(topicPartition);
+                                addFuture.complete(null);
+                                return;
+                            }
+                            if (!address.isPresent()) {
+                                log.warn("No address for broker for topic partition {}", topicPartition);
+                                unknownBrokerTopicList.add(topicPartition);
+                                addFuture.complete(null);
+                                return;
+                            }
+                            addressAndPartitionMap.compute(address.get(), (__, set) -> {
+                                if (set == null) {
+                                    set = new ArrayList<>();
+                                }
+                                set.add(topicPartition);
+                                return set;
+                            });
+                            addFuture.complete(null);
+                        });
+                    });
         }
         FutureUtil.waitForAll(addressFutureList).whenComplete((ignored, throwable) -> {
             addressAndPartitionMap.forEach((address, partitions) -> {

--- a/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMarkerChannelManager.java
+++ b/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMarkerChannelManager.java
@@ -54,6 +54,7 @@ import org.apache.kafka.common.requests.WriteTxnMarkersRequest;
 import org.apache.kafka.common.requests.WriteTxnMarkersRequest.TxnMarkerEntry;
 import org.apache.pulsar.client.api.Authentication;
 import org.apache.pulsar.client.impl.AuthenticationUtil;
+import org.apache.pulsar.common.naming.TopicName;
 import org.apache.pulsar.common.util.FutureUtil;
 import org.apache.pulsar.common.util.netty.ChannelFutures;
 import org.eclipse.jetty.util.BlockingArrayQueue;
@@ -323,7 +324,7 @@ public class TransactionMarkerChannelManager {
                         }
 
                         CompletableFuture<Optional<InetSocketAddress>> addressFuture =
-                                kopBrokerLookupManager.findBroker(pulsarTopic, sslEndPoint);
+                                kopBrokerLookupManager.findBroker(TopicName.get(pulsarTopic), sslEndPoint);
 
                         addressFuture.whenComplete((address, throwable) -> {
                             if (throwable != null) {

--- a/tests/src/test/java/io/streamnative/pulsar/handlers/kop/MetricsProviderTest.java
+++ b/tests/src/test/java/io/streamnative/pulsar/handlers/kop/MetricsProviderTest.java
@@ -67,7 +67,7 @@ public class MetricsProviderTest extends KopProtocolHandlerTestBase {
         this.conf.setOffsetsTopicNumPartitions(50);
         this.conf.setKafkaTxnLogTopicNumPartitions(50);
         this.conf.setKafkaTransactionCoordinatorEnabled(true);
-        this.conf.setBrokerDeduplicationEnabled(true);
+        this.conf.setBrokerDeduplicationEnabled(false);
         super.internalSetup();
         log.info("success internal setup");
     }

--- a/tests/src/test/java/io/streamnative/pulsar/handlers/kop/TransactionTest.java
+++ b/tests/src/test/java/io/streamnative/pulsar/handlers/kop/TransactionTest.java
@@ -306,10 +306,8 @@ public class TransactionTest extends KopProtocolHandlerTestBase {
     }
 
     private KafkaProducer<Integer, String> buildTransactionProducer(String transactionalId) {
-        String kafkaServer = "localhost:" + getKafkaBrokerPort();
-
         Properties producerProps = new Properties();
-        producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaServer);
+        producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, getKafkaServerAdder());
         producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, IntegerSerializer.class.getName());
         producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
         producerProps.put(ProducerConfig.REQUEST_TIMEOUT_MS_CONFIG, 1000 * 10);
@@ -320,10 +318,8 @@ public class TransactionTest extends KopProtocolHandlerTestBase {
     }
 
     private KafkaConsumer<Integer, String> buildTransactionConsumer(String groupId, String isolation) {
-        String kafkaServer = "localhost:" + getKafkaBrokerPort();
-
         Properties consumerProps = new Properties();
-        consumerProps.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaServer);
+        consumerProps.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, getKafkaServerAdder());
         consumerProps.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, IntegerDeserializer.class.getName());
         consumerProps.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
         consumerProps.put(ConsumerConfig.REQUEST_TIMEOUT_MS_CONFIG, 1000 * 10);
@@ -336,10 +332,8 @@ public class TransactionTest extends KopProtocolHandlerTestBase {
     }
 
     private KafkaProducer<Integer, String> buildIdempotenceProducer() {
-        String kafkaServer = "localhost:" + getKafkaBrokerPort();
-
         Properties producerProps = new Properties();
-        producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaServer);
+        producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, getKafkaServerAdder());
         producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, IntegerSerializer.class.getName());
         producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
         producerProps.put(ProducerConfig.REQUEST_TIMEOUT_MS_CONFIG, 1000 * 10);
@@ -347,6 +341,13 @@ public class TransactionTest extends KopProtocolHandlerTestBase {
 
         addCustomizeProps(producerProps);
         return new KafkaProducer<>(producerProps);
+    }
+
+    /**
+     * Get the Kafka server address.
+     */
+    private String getKafkaServerAdder() {
+        return "localhost:" + getKafkaBrokerPort();
     }
 
     protected void addCustomizeProps(Properties producerProps) {

--- a/tests/src/test/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMarkerChannelManagerTest.java
+++ b/tests/src/test/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMarkerChannelManagerTest.java
@@ -1,0 +1,133 @@
+/**
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.streamnative.pulsar.handlers.kop.coordinator.transaction;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.when;
+import static org.testng.Assert.assertEquals;
+
+import com.google.common.collect.Sets;
+import io.streamnative.pulsar.handlers.kop.KafkaServiceConfiguration;
+import io.streamnative.pulsar.handlers.kop.KopBrokerLookupManager;
+import io.streamnative.pulsar.handlers.kop.utils.timer.MockTime;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import org.apache.bookkeeper.common.util.OrderedScheduler;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.record.RecordBatch;
+import org.apache.kafka.common.requests.TransactionResult;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+/**
+ * Unit test {@link TransactionMarkerChannelManager}.
+ */
+public class TransactionMarkerChannelManagerTest {
+
+    private TransactionMarkerChannelManager transactionMarkerChannelManager;
+
+    private KopBrokerLookupManager kopBrokerLookupManager;
+
+    private TransactionStateManager txnStateManager;
+
+    private OrderedScheduler scheduler;
+    private final MockTime time = new MockTime();
+    private final Set<TopicPartition> partitions = Sets.newHashSet(new TopicPartition("topic1", 0));
+
+    @BeforeMethod
+    public void setUp() throws Exception {
+        txnStateManager = mock(TransactionStateManager.class);
+        when(txnStateManager.partitionFor(any())).thenReturn(0);
+
+        kopBrokerLookupManager = mock(KopBrokerLookupManager.class);
+        scheduler = mock(OrderedScheduler.class);
+        KafkaServiceConfiguration conf = new KafkaServiceConfiguration();
+        transactionMarkerChannelManager = spy(new TransactionMarkerChannelManager(
+                "public",
+                conf,
+                txnStateManager,
+                kopBrokerLookupManager,
+                false,
+                "public/default",
+                scheduler
+                ));
+    }
+
+    @AfterMethod
+    public void tearDown() {
+        transactionMarkerChannelManager = null;
+        kopBrokerLookupManager = null;
+        scheduler.shutdown();
+        scheduler = null;
+        txnStateManager = null;
+    }
+
+    @DataProvider
+    public static Object[] isTopicExistsList() {
+        return new Object[] {true, false};
+    }
+
+    @Test(dataProvider = "isTopicExistsList")
+    public void testTopicDeletedBeforeWriteMarker(boolean isTopicExists) {
+        when(kopBrokerLookupManager.isTopicExists(any())).thenReturn(CompletableFuture.completedFuture(isTopicExists));
+        String transactionalId = "known";
+        long producerId = 10L;
+        short producerEpoch = 1;
+        int txnTimeoutMs = 1;
+        TransactionMetadata txnMetadata = TransactionMetadata.builder()
+                .transactionalId(transactionalId)
+                .producerId(producerId)
+                .lastProducerId(producerId)
+                .producerEpoch(producerEpoch)
+                .lastProducerEpoch(RecordBatch.NO_PRODUCER_EPOCH)
+                .txnTimeoutMs(txnTimeoutMs)
+                .state(TransactionState.PREPARE_COMMIT)
+                .topicPartitions(partitions)
+                .txnStartTimestamp(time.milliseconds())
+                .txnLastUpdateTimestamp(time.milliseconds())
+                .build();
+        TransactionMetadata.TxnTransitMetadata transition = TransactionMetadata.TxnTransitMetadata.builder()
+                .producerId(producerId)
+                .lastProducerId(producerId)
+                .producerEpoch(producerEpoch)
+                .lastProducerEpoch(RecordBatch.NO_PRODUCER_EPOCH)
+                .txnTimeoutMs(txnTimeoutMs)
+                .txnState(TransactionState.COMPLETE_COMMIT)
+                .topicPartitions(partitions)
+                .txnStartTimestamp(time.milliseconds())
+                .txnLastUpdateTimestamp(time.milliseconds())
+                .build();
+
+        int coordinatorEpoch = 0;
+        TransactionStateManager.CoordinatorEpochAndTxnMetadata epochAndTxnMetadata =
+                new TransactionStateManager.CoordinatorEpochAndTxnMetadata(coordinatorEpoch, txnMetadata);
+        when(txnStateManager.getTransactionState(transactionalId))
+                .thenReturn(new ErrorsAndData<>(Optional.of(epochAndTxnMetadata)));
+        transactionMarkerChannelManager.addTxnMarkersToSend(
+                coordinatorEpoch,
+                TransactionResult.COMMIT,
+                txnMetadata,
+                transition,
+                "public/default"
+                );
+
+        assertEquals(transactionMarkerChannelManager.getTransactionsWithPendingMarkers().isEmpty(), !isTopicExists);
+    }
+
+}

